operator/internal/decommissioning/statefulset_decomissioner.go

```diff
@@ -0,0 +1,753 @@
+		opt(decommissioner)
+	}
+
+	decommissioner.delayedBrokerIDCache = NewCategorizedDelayedCache[types.NamespacedName, int](decommissioner.delayedCacheMaxCount, decommissioner.delayedCacheInterval)
+	decommissioner.delayedVolumeCache = NewCategorizedDelayedCache[types.NamespacedName, types.NamespacedName](decommissioner.delayedCacheMaxCount, decommissioner.delayedCacheInterval)
+
+	return decommissioner
+}
+
+// +kubebuilder:rbac:groups=apps,resources=statefulsets,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;delete
+// +kubebuilder:rbac:groups=core,resources=persistentvolumes,verbs=patch
+// +kubebuilder:rbac:groups=core,resources=events,verbs=create;patch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch
+
+// +kubebuilder:rbac:groups=apps,namespace=default,resources=statefulsets,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,namespace=default,resources=pods,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,namespace=default,resources=persistentvolumeclaims,verbs=get;list;watch;delete
+// +kubebuilder:rbac:groups=core,namespace=default,resources=persistentvolumes,verbs=patch
+// +kubebuilder:rbac:groups=core,namespace=default,resources=events,verbs=create;patch
+// +kubebuilder:rbac:groups=core,namespace=default,resources=secrets,verbs=get;list;watch
+
+func (s *StatefulSetDecomissioner) SetupWithManager(mgr ctrl.Manager) error {
+	pvcPredicate, err := predicate.LabelSelectorPredicate(
+		metav1.LabelSelector{
+			MatchExpressions: []metav1.LabelSelectorRequirement{{
+				Key:      k8sNameLabelKey, // look for only redpanda owned pvcs
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   []string{"redpanda"},
+			}, {
+				Key:      k8sComponentLabelKey, // make sure the PVC is part of the statefulset
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   []string{"redpanda-statefulset"},
+			}, {
+				Key:      k8sInstanceLabelKey, // make sure we have a cluster name
+				Operator: metav1.LabelSelectorOpExists,
+			}},
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.StatefulSet{}).
+		Owns(&corev1.Pod{}).
+		// PVCs don't have a "true" owner ref, so instead we attempt to map backwards via labels
+		Watches(&corev1.PersistentVolumeClaim{}, handler.EnqueueRequestsFromMapFunc(func(_ context.Context, o client.Object) []ctrl.Request {
+			claim := o.(*corev1.PersistentVolumeClaim)
+			labels := claim.GetLabels()
+
+			// a bit of defensive programming, but we should always have labels due to our use
+			// of a predicate
+			if labels == nil {
+				// we have no labels, so we can't map anything
+				return nil
+			}
+
+			release := labels[k8sInstanceLabelKey]
+			if release == "" {
+				// we have an invalid release name, so skip
+				return nil
+			}
+
+			if !strings.HasPrefix(claim.Name, datadirVolume+"-") {
```

Comment:
Ah, I think that's a good call-out that we should likely handle cleaning up the PVCs that we're optionally creating as well. The original code was based purely off of what the decommission controller was doing previously, which was for just `datadir`, but I think it'd make sense to check for any of our optionally provisioned PVCs as well.

Reply:
Think I'm probably just going to leave this as-is for now since we don't try and decommission PVCs for non-datadir right now anyway. But I think it'd make sense to do so as a future iteration.